## House Prices - Advanced Regression Techniques

## notes
target = SalePrice  

problems: too many features...

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('./train.csv')

In [3]:
df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [4]:
df.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000  ...   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726  ...   
std       1.112799    30.202904     20.645407   181.066207   456.098091  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000  ...   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000   
mean     94.244521    46.660274      21.954110     3.409589    15.060959   
std     125.338794    66.256028      61.119149    29.317331    55.757415   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    25.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000  
mean      2.758904     43.489041     6.321918  2007.815753  180921.195890  
std      40.177307    496.123024     2.703626     1.328095   79442.502883  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129975.000000  
50%       0.000000      0.000000     6.000000  2008.000000  163000.000000  
75%       0.000000      0.000000     8.000000  2009.000000  214000.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

## Handling the missing values

In [6]:
print(df.isnull().sum().sort_values(ascending=False).head(20))

PoolQC          1453
MiscFeature     1406
Alley           1369
Fence           1179
MasVnrType       872
FireplaceQu      690
LotFrontage      259
GarageYrBlt       81
GarageCond        81
GarageType        81
GarageFinish      81
GarageQual        81
BsmtFinType2      38
BsmtExposure      38
BsmtQual          37
BsmtCond          37
BsmtFinType1      37
MasVnrArea         8
Electrical         1
Id                 0
dtype: int64


In [7]:
print(f'{df.isnull().sum().sort_values(ascending=False).head(20).sum()} missing values')
cat_features = df.select_dtypes(include=['object']).shape[1]
print(f"{cat_features} categorical features")
num_features = df.select_dtypes(include=['int64', 'float64']).shape[1]
print(f"{num_features} numerical features")

7829 missing values
43 categorical features
38 numerical features


In [8]:
df.shape

(1460, 81)

## Reports
We have 1460 enteries and 81 features 
7829 missing values  
43 categorical features  
38 numerical features  
whats scary? too many features and too many missing values.

## Training a linear model, without the missing values

In [9]:
# Select only numerical columns
df_numeric = df.select_dtypes(include=['int64', 'float64'])

# Drop columns with any missing values
df_clean = df_numeric.dropna(axis=0)

print(f"Original shape: {df.shape}")
print(f"Numeric shape: {df_numeric.shape}")
print(f"Clean shape: {df_clean.shape}")

Original shape: (1460, 81)
Numeric shape: (1460, 38)
Clean shape: (1121, 38)


Original shape: (1460, 81)   
Numeric shape: (1460, 38)  
Clean shape: (1460, 35)  

## Spliting the data

In [20]:
from sklearn.model_selection import train_test_split, cross_val_score

In [11]:
X = df_clean.drop(['SalePrice', 'Id'], axis=1)
y = df_clean['SalePrice']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## Training the model

In [12]:
from sklearn.linear_model import LinearRegression

In [13]:
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

In [14]:
coef_df = pd.DataFrame({
    'Feature': X.columns,
    'Coefficient': model.coef_
})
coef_df = coef_df.sort_values('Coefficient', ascending=False).reset_index(drop=True)
coef_df

Feature   Coefficient
0     OverallQual  19654.291953
1      GarageCars  18344.552805
2    BsmtFullBath  11565.128964
3     OverallCond   5072.289462
4        FullBath   5010.143758
5      Fireplaces   4793.918236
6    TotRmsAbvGrd   4587.157867
7    BsmtHalfBath   1338.748402
8       YearBuilt    286.898275
9    YearRemodAdd    125.155507
10    ScreenPorch     66.973618
11      3SsnPorch     58.657103
12      GrLivArea     29.344522
13     MasVnrArea     26.492720
14     WoodDeckSF     20.256803
15       2ndFlrSF     13.903081
16  EnclosedPorch     11.847823
17       1stFlrSF     10.350316
18    TotalBsmtSF      6.237704
19     GarageArea      5.860430
20   LowQualFinSF      5.091125
21     BsmtFinSF1      4.750186
22     BsmtFinSF2      4.490713
23        LotArea      0.429203
24      BsmtUnfSF     -3.003196
25    OpenPorchSF     -3.143833
26        MiscVal     -3.910092
27         MoSold     -5.280061
28    GarageYrBlt    -17.382247
29       PoolArea    -61.926903
30    LotFrontage    -62.224525
31         YrSold   -191.225833
32     MSSubClass   -209.840235
33       HalfBath  -1955.444800
34   BedroomAbvGr  -9486.849775
35   KitchenAbvGr -21475.023726

## This is just to see the impact of the dropped features with missing value on the model

In [18]:
# Get the features from both tables
missing_features = df.isnull().sum().sort_values(ascending=False).head(20).index.tolist()
coefficient_features = coef_df['Feature'].tolist()

# Find common features
common_features = set(missing_features).intersection(set(coefficient_features))

# Get their positions in the coefficient table
for feature in common_features:
    rank = coefficient_features.index(feature) + 1  # +1 for human-readable rank
    coefficient = coef_df.loc[coef_df['Feature'] == feature, 'Coefficient'].values[0]
    print(f"'{feature}' is at rank #{rank} in coefficients with value: {coefficient:.2f}")

# Summary
print(f"\nTotal common features: {len(common_features)} out of {len(coefficient_features)}")

'MasVnrArea' is at rank #14 in coefficients with value: 26.49
'LotFrontage' is at rank #31 in coefficients with value: -62.22
'GarageYrBlt' is at rank #29 in coefficients with value: -17.38

Total common features: 3 out of 36


In [15]:
y_pred = model.predict(X_test)

In [21]:
cv_scores = cross_val_score(
    model,
    X,
    y,
    scoring="neg_root_mean_squared_error",
    cv=5
)


In [22]:
rmse_scores = -cv_scores

In [23]:
print("RMSE scores:", rmse_scores)
print("Mean RMSE:", rmse_scores.mean())
print("Std RMSE:", rmse_scores.std())

RMSE scores: [29587.07709133 37665.99103449 40100.65600937 32378.41339688
 54086.83339133]
Mean RMSE: 38763.79418468276
Std RMSE: 8517.900283988969


## reports
Well CV tells me that the model is usually around ~38$ off the actual value. also it tells me that the results must be very noisy.  